okta/okta.go

```diff
@@ -819,21 +808,26 @@ func (m *userManager) getTokenInfo(ctx context.Context, token, tokenType string)
 	if err != nil {
 		return nil, errors.Wrap(err, "request to introspect token returned error")
 	}
+	tokenInfo := &introspectResponse{}
+	return tokenInfo, readResp(resp, tokenInfo)
+}
+
+// readResp verifies that an Okta response is OK and reads an Okta response body
+// in JSON format to the given output out.
+func readResp(resp *http.Response, out error) error {
+	catcher := grip.NewBasicCatcher()
 	if resp.StatusCode != http.StatusOK {
-		catcher := grip.NewBasicCatcher()
-		catcher.Errorf("received unexpected status code %d", resp.StatusCode)
-		catcher.Wrap(resp.Body.Close(), "error closing response body")
-		return nil, catcher.Resolve()
+		catcher.Errorf("unexpected status code %d", resp.StatusCode)
 	}
-
-	tokenInfo := &introspectResponse{}
-	if err := gimlet.GetJSONUnlimited(resp.Body, tokenInfo); err != nil {
-		return nil, errors.WithStack(err)
+	if err := gimlet.GetJSONUnlimited(resp.Body, out); err != nil {
```

Comment:
This will change out won't it? Would it be better to just return the parsed body and let the caller choose how to use it?

Reply:
This one's kind of weird, but it's supposed to unmarshal into `out`. All Okta server responses return a JSON body with an `error` and `error_description` after receiving a non-200 response, so I made it so that the Okta server responses implement the `error` interface and check the special Okta error below this. In the case of success, `out` will still contain the successful fields (e.g. for `tokenResponse`, `out` will return an error if `error`/`error_description` are set, but if not, it'll set the actual payload).